trezor/trezor-suite
# @trezor/transport refactor

### Based on
- [x] #7321
- [x]  https://github.com/trezor/trezor-suite/pull/7411
- [x] https://github.com/trezor/trezor-suite/pull/7409
- [x] https://github.com/trezor/trezor-suite/pull/7336
- [x] https://github.com/trezor/trezor-suite/pull/7346
- [x] https://github.com/trezor/trezor-suite/pull/7383
- [x] https://github.com/trezor/trezor-suite/pull/7411
- [x] https://github.com/trezor/trezor-suite/pull/7443
- [x] #7821
- [x] #7870, #7844, #7857 needed to updated jest and run tests with AbortController in connect
- [x] #7922 needed to run make AbortController in tests work
 
Changes in @trezor/transport: 
- removed debug-link support all together. We have not used it since trezor-user-env came into service.
- sessions management. I am adding a module that is called "sessions background" it is used to control access to usb/device and holds information about sessions. This is not new, previously it was implemented [here](https://github.dev/trezor/trezor-suite/blob/develop/packages/transport/src/lowlevel/sharedConnectionWorker.ts#L150).  This module could be used in various contexts. At the moment, we only have [SharedWorker](https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker). 
- created `interfaces` folder with lowest-possible-level unifying abstraction over native transport interfaces. so far only webusb.
- added abstract transport class for API unification which itself inhertis from a strongly typed event emiter. no method from transport should throw error. each method should return either `Success` or `Error`
- added abstract USB class, derived from abstract transport class. Used for code sharing between webusb and nodeusb. This is basically compilation of [withSharedConnections.ts](https://github.dev/trezor/trezor-suite/blob/develop/packages/transport/src/lowlevel/withSharedConnections.ts) and [webusb.ts](https://github.dev/trezor/trezor-suite/blob/55f7ea4f0f0c544e300dc1d8a080d558404ae876/packages/transport/src/lowlevel/webusb.ts). This is the place that implements low-level transport interface and sessions.
- webusb now does not run iteration cycle in 500ms endless loop. The reasons are that it does not feel right if there are connection events available and also it probably would not work for nodeusb where enumaration cycle takes considerably longer than 500ms 
- all transport calls are now abortable
- race conditions related to "cancel" method should now be fixed. Errors such as "Unexpected response, expected Address, received Featrues" should not appear anymore or at least should appear less often.

### Related issues
- resolves #5036
- resolves #5035 by moving former DescriptorStream logic into the abstract transport class. 
- resolves #5336 because these files will not exist anymore and there will be only transports that inherit from abstract transport class
- resolves #5597
- resolves #4268
- closes #4601 - problems described there do not seem relevant anymore
- related #5332
- related #4041
- enables #4770
- #5068 - should at least not be unhandled now.
- #5069 - should at least not be unhandeld now and might even be fixed. This seems to be part of cancelling problem, here trying to read/write from device that has been released by bridge.
- related #6886 - connect now accepts transport as a param in init method.
- please retest #5543
- please retest #4670
- please retest #4267
- #5056 would be interesting to check/finish this test

## QA testing tips:
In the first place, focus on regression in communication with device. Nodeusb feature is still hidden behind debug (and I am even considering pulling it out into separate PR based on this). Some tips what to test:
- `google-chrome http://localhost:8000/ http://localhost:8000/` open multiple tabs at once
- debugging shared workers [chrome://inspect/#workers](chrome://inspect/#workers)
- multiple devices connected
- mad-reconnecting in the middle of communication with device
- cancel message (pin, passphrase, etc)
- all of above for both webusb and bridge
- when using bridge transport, app should survive when computer goes to sleep 

### QA
- [first smoke test](https://www.notion.so/satoshilabs/Test-connect-branch-before-merge-d72cc3ebffb94771b46a25ab6ed0997c)
- [test 2](https://www.notion.so/satoshilabs/Transport-refactor-final-retest-b3e498a0a87a4c77bfd5a07e6ba53a59?pvs=4)
